core/src/main/java/jenkins/slaves/restarter/UnixSlaveRestarter.java

```diff
@@ -57,11 +59,61 @@ public void restart() throws Exception {
         }
 
         // exec to self
-        String exe = Daemon.getCurrentExecutable();
+        String exe = getCurrentExecutable();
         LIBC.execv(exe, new StringArray(args.toArray(new String[0])));
         throw new IOException("Failed to exec '" + exe + "' " + LIBC.strerror(Native.getLastError()));
     }
 
+    /**
+     * Gets the current executable name.
+     */
+    private static String getCurrentExecutable() {
+        int pid = LIBC.getpid();
+        String name = "/proc/" + pid + "/exe";
+        File exe = new File(name);
+        if (exe.exists()) {
+            try {
+                String path = resolveSymlink(exe);
+                if (path != null) {
+                    return path;
+                }
+            } catch (IOException e) {
+                LOGGER.log(FINE, "Failed to resolve symlink " + exe, e);
+            }
+            return name;
+        }
+
+        // cross-platform fallback
+        return System.getProperty("java.home") + "/bin/java";
+    }
+
+    private static String resolveSymlink(File link) throws IOException {
+        String filename = link.getAbsolutePath();
```

Comment:
if this is just resolving symlinks (`/symllink/to/a` -> `/foo/bar/to/a`) (as the function suggests) then can we remove this code and replace it with just `link.getCanonicalPath()`?
Windows has funny symbolic type links which are not symlinks - but we are heavily in the realm of Unix here.

Reply:
I didn't write this code. I just moved it. I am not going to change code that is just being moved. If you feel that it can be improved, feel free to open a separate PR.